File: sdk/src/types/agent.ts
export interface AgentSocialLinks {
  website?: string;
  twitter?: string;
  github?: string;
  linkedin?: string;
}

/** `unlisted` is reachable by direct link/id but excluded from public discovery listings. */
export type AgentVisibility = 'private' | 'unlisted' | 'public';
export type AgentCategory =
  'productivity' | 'coding' | 'creative' | 'research' | 'roleplay' | 'other';

/**
 * Note: like `Skill`, this mirrors the real wire shape (`_id`, raw
 * domain/ownerType fields) — `getOne()` returns `skills`/`mcps`/
 * `knowledgeBases` populated as objects, while `create()`/`update()`/
 * `list()` return them as bare id strings; typed loosely here to reflect
 * that real difference rather than picking one shape and being wrong for
 * the other calls.
 */
export interface Agent {
  _id: string;
  domain: string;
  ownerType: 'PersonaUser' | 'Project' | 'ExternalUser';
  ownerId?: string;
  externalOwnerId?: string;
  name: string;
  /** URL-safe, unique within the Domain; used in some public-facing routes. */
  slug: string;
  description?: string;
  avatar?: string;
  tags?: string[];
  /** Short one-liner shown in list/card views. */
  tagline?: string;
  /** Longer free-text bio shown on the Agent's own profile view. */
  bio?: string;
  personalityTraits?: string[];
  socialLinks?: AgentSocialLinks;
  /** Stripped from the response when the calling identity doesn't own this Agent. */
  systemPrompt?: string;
  /** Stripped from the response when the calling identity doesn't own this Agent. */
  providerId?: string;
  /** Overrides the referenced Provider's `defaultModel` when set. */
  modelName?: string;
  webSearchEnabled: boolean;
  visibility: AgentVisibility;
  category: AgentCategory;
  /** Bare id strings on `create`/`update`/`list`; populated objects on `get()`. */
  skills?: unknown[];
  /** Bare id strings on `create`/`update`/`list`; populated objects on `get()`. */
  mcps?: unknown[];
  /** Bare id strings on `create`/`update`/`list`; populated objects on `get()`. */
  knowledgeBases?: unknown[];
  isActive: boolean;
  /** Whether this is the Project's designated default/primary Agent. */
  isMainAgent: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface CreateAgentInput {
  name: string;
  /** The instructions that define this Agent's behavior/persona. */
  systemPrompt: string;
  /** Must reference a Provider already created via `providers.create()`. */
  providerId: string;
  description?: string;
  avatar?: string;
  tags?: string[];
  tagline?: string;
  bio?: string;
  personalityTraits?: string[];
  socialLinks?: AgentSocialLinks;
  /** Overrides the referenced Provider's `defaultModel`. */
  modelName?: string;
  /** @default false */
  webSearchEnabled?: boolean;
  /** @default 'private' */
  visibility?: AgentVisibility;
  /** @default 'other' */
  category?: AgentCategory;
  /** Skill ids to attach at creation time. */
  skills?: string[];
  /** MCP server ids to attach at creation time. */
  mcps?: string[];
  /** Knowledge base ids to attach at creation time. */
  knowledgeBases?: string[];
  /** @default true */
  isActive?: boolean;
}

/** All fields optional — only what you pass is changed. */
export interface UpdateAgentInput {
  name?: string;
  description?: string;
  avatar?: string;
  tags?: string[];
  tagline?: string;
  bio?: string;
  personalityTraits?: string[];
  socialLinks?: AgentSocialLinks;
  systemPrompt?: string;
  providerId?: string;
  modelName?: string;
  webSearchEnabled?: boolean;
  /** Replaces the entire array — this is not a merge/append. */
  skills?: string[];
  /** Replaces the entire array — this is not a merge/append. */
  mcps?: string[];
  /** Replaces the entire array — this is not a merge/append. */
  knowledgeBases?: string[];
  visibility?: AgentVisibility;
  category?: AgentCategory;
  isActive?: boolean;
}

export interface DiscoverAgentsParams {
  /** @default 1 */
  page?: number;
  /** @default 20 */
  limit?: number;
  /** Free-text match against `name`/`description`/`tagline`. */
  search?: string;
  category?: AgentCategory;
  /** Restricts to the asserted external user's own Agents (`ProjectRuntimeContext` only). */
  scope?: 'mine';
}

File: sdk-python/src/personaai/types/agent.py
"""Mirrors ``sdk/src/types/agent.ts``."""

from __future__ import annotations

from typing import Literal, TypedDict


class AgentSocialLinks(TypedDict, total=False):
    website: str
    twitter: str
    github: str
    linkedin: str


# "unlisted" is reachable by direct link/id but excluded from public discovery listings.
AgentVisibility = Literal["private", "unlisted", "public"]
AgentCategory = Literal["productivity", "coding", "creative", "research", "roleplay", "other"]


class _AgentRequired(TypedDict):
    _id: str
    domain: str
    ownerType: Literal["PersonaUser", "Project", "ExternalUser"]
    name: str
    slug: str  # URL-safe, unique within the Domain; used in some public-facing routes
    webSearchEnabled: bool
    visibility: AgentVisibility
    category: AgentCategory
    isActive: bool
    isMainAgent: bool  # whether this is the Project's designated default/primary Agent
    createdAt: str
    updatedAt: str


class Agent(_AgentRequired, total=False):
    """Note: like ``Skill``, this mirrors the real wire shape (``_id``, raw
    domain/ownerType fields) — ``get()`` returns ``skills``/``mcps``/
    ``knowledgeBases`` populated as objects, while ``create()``/``update()``/
    ``list()`` return them as bare id strings; typed loosely here
    (``list[object]``) to reflect that real difference rather than picking
    one shape and being wrong for the other calls."""

    ownerId: str
    externalOwnerId: str
    description: str
    avatar: str
    tags: list[str]
    tagline: str  # short one-liner shown in list/card views
    bio: str  # longer free-text bio shown on the Agent's own profile view
    personalityTraits: list[str]
    socialLinks: AgentSocialLinks
    systemPrompt: str  # stripped from the response when the caller doesn't own this Agent
    providerId: str  # stripped from the response when the caller doesn't own this Agent
    modelName: str  # overrides the referenced Provider's defaultModel when set
    skills: list[object]  # bare id strings on create/update/list; populated objects on get()
    mcps: list[object]  # bare id strings on create/update/list; populated objects on get()
    knowledgeBases: list[
        object
    ]  # bare id strings on create/update/list; populated objects on get()


class _CreateAgentInputRequired(TypedDict):
    name: str
    systemPrompt: str  # the instructions that define this Agent's behavior/persona
    providerId: str  # must reference a Provider already created via providers.create()


class CreateAgentInput(_CreateAgentInputRequired, total=False):
    description: str
    avatar: str
    tags: list[str]
    tagline: str
    bio: str
    personalityTraits: list[str]
    socialLinks: AgentSocialLinks
    modelName: str  # overrides the referenced Provider's defaultModel
    webSearchEnabled: bool  # default: False
    visibility: AgentVisibility  # default: 'private'
    category: AgentCategory  # default: 'other'
    skills: list[str]  # Skill ids to attach at creation time
    mcps: list[str]  # MCP server ids to attach at creation time
    knowledgeBases: list[str]  # Knowledge base ids to attach at creation time
    isActive: bool  # default: True


class UpdateAgentInput(TypedDict, total=False):
    """All fields optional — only what you pass is changed."""

    name: str
    description: str
    avatar: str
    tags: list[str]
    tagline: str
    bio: str
    personalityTraits: list[str]
    socialLinks: AgentSocialLinks
    systemPrompt: str
    providerId: str
    modelName: str
    webSearchEnabled: bool
    skills: list[str]  # replaces the entire list — this is not a merge/append
    mcps: list[str]  # replaces the entire list — this is not a merge/append
    knowledgeBases: list[str]  # replaces the entire list — this is not a merge/append
    visibility: AgentVisibility
    category: AgentCategory
    isActive: bool


class DiscoverAgentsParams(TypedDict, total=False):
    page: int  # default: 1
    limit: int  # default: 20
    search: str  # free-text match against name/description/tagline
    category: AgentCategory
    scope: Literal["mine"]  # restricts to the asserted external user's own Agents (runtime-only)
